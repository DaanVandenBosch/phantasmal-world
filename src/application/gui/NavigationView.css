.application_NavigationView {
    box-sizing: border-box;
    display: flex;
    flex-direction: row;
    align-items: stretch;
    background-color: hsl(0, 0%, 10%);
    border-bottom: solid 2px var(--bg-color);
}

.application_NavigationView_spacer {
    flex: 1;
}

.application_NavigationView_server {
    display: flex;
    align-items: center;
}

.application_NavigationView_server > * {
    margin: 0 2px;
}

.application_NavigationView_github {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    width: 30px;
    font-size: 16px;
    color: var(--control-text-color);
}

.application_NavigationView_github:hover {
    color: var(--control-text-color-hover);
}
