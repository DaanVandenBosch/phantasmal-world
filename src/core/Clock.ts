export interface Clock {
    now(): Date;
}

export class DateClock implements Clock {
    now(): Date {
        return new Date();
    }
}
