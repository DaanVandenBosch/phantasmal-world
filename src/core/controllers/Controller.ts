import { Disposable } from "../observable/Disposable";
import { Disposer } from "../observable/Disposer";

export abstract class Controller implements Disposable {
    private readonly disposer = new Disposer();

    dispose(): void {
        this.disposer.dispose();
    }

    protected disposable<T extends Disposable>(disposable: T): T {
        return this.disposer.add(disposable);
    }

    protected disposables(...disposables: Disposable[]): void {
        this.disposer.add_all(...disposables);
    }
}
