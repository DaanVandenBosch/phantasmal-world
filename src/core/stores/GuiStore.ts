import { WritableProperty } from "../observable/property/WritableProperty";
import { Disposable } from "../observable/Disposable";
import { property } from "../observable";

export enum GuiTool {
    Viewer,
    QuestEditor,
    HuntOptimizer,
}

const GUI_TOOL_TO_STRING = new Map([
    [GuiTool.Viewer, "viewer"],
    [GuiTool.QuestEditor, "quest_editor"],
    [GuiTool.HuntOptimizer, "hunt_optimizer"],
]);
const STRING_TO_GUI_TOOL = new Map([...GUI_TOOL_TO_STRING.entries()].map(([k, v]) => [v, k]));

class GuiStore implements Disposable {
    readonly tool: WritableProperty<GuiTool> = property(GuiTool.Viewer);

    private hash_disposer = this.tool.observe(({ value: tool }) => {
        window.location.hash = `#/${gui_tool_to_string(tool)}`;
    });

    constructor() {
        const tool = window.location.hash.slice(2);
        this.tool.val = string_to_gui_tool(tool) || GuiTool.Viewer;
    }

    dispose(): void {
        this.hash_disposer.dispose();
    }
}

export const gui_store = new GuiStore();

function string_to_gui_tool(tool: string): GuiTool | undefined {
    return STRING_TO_GUI_TOOL.get(tool);
}

function gui_tool_to_string(tool: GuiTool): string {
    const str = GUI_TOOL_TO_STRING.get(tool);

    if (str) {
        return str;
    } else {
        throw new Error(`To string not implemented for ${(GuiTool as any)[tool]}.`);
    }
}
