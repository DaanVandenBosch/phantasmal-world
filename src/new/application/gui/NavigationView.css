.application_NavigationView {
    box-sizing: border-box;
    display: flex;
    flex-direction: row;
    align-items: stretch;
    background-color: hsl(0, 0%, 12%);
    border-bottom: solid 2px var(--bg-color);
}

.application_ToolButton input {
    display: none;
}

.application_ToolButton label {
    box-sizing: border-box;
    display: inline-block;
    font-size: 16px;
    height: 100%;
    padding: 0 20px;
    line-height: 40px;
}

.application_ToolButton label:hover {
    color: hsl(200, 25%, 85%);
    background-color: hsl(0, 0%, 16%);
}

.application_ToolButton input:checked + label {
    color: hsl(200, 50%, 85%);
    background-color: hsl(0, 0%, 20%);
}
