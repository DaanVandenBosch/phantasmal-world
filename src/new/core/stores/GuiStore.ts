import { Observable } from "../observable/Observable";

export enum GuiTool {
    Viewer,
    QuestEditor,
    HuntOptimizer,
}

const GUI_TOOL_TO_STRING = new Map([
    [GuiTool.Viewer, "viewer"],
    [GuiTool.QuestEditor, "quest_editor"],
    [GuiTool.HuntOptimizer, "hunt_optimizer"],
]);
const STRING_TO_GUI_TOOL = new Map([...GUI_TOOL_TO_STRING.entries()].map(([k, v]) => [v, k]));

class GuiStore {
    tool_prop = new Observable<GuiTool>(GuiTool.Viewer);

    get tool(): GuiTool {
        return this.tool_prop.get();
    }

    set tool(tool: GuiTool) {
        window.location.hash = `#/${gui_tool_to_string(tool)}`;
        this.tool_prop.set(tool);
    }

    constructor() {
        const tool = window.location.hash.slice(2);
        this.tool = string_to_gui_tool(tool) || GuiTool.Viewer;
    }
}

export const gui_store = new GuiStore();

function string_to_gui_tool(tool: string): GuiTool | undefined {
    return STRING_TO_GUI_TOOL.get(tool);
}

function gui_tool_to_string(tool: GuiTool): string {
    const str = GUI_TOOL_TO_STRING.get(tool);

    if (str) {
        return str;
    } else {
        throw new Error(`To string not implemented for ${(GuiTool as any)[tool]}.`);
    }
}
