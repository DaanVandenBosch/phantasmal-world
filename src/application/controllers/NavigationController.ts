import { Controller } from "../../core/controllers/Controller";
import { Property } from "../../core/observable/property/Property";
import { GuiStore, GuiTool } from "../../core/stores/GuiStore";
import { property } from "../../core/observable";
import { Clock } from "../../core/Clock";

export class NavigationController extends Controller {
    private readonly _internet_time = property("@");
    private readonly internet_time_interval: any;

    readonly tool: Property<GuiTool>;
    readonly internet_time: Property<string> = this._internet_time;

    constructor(private readonly gui_store: GuiStore, private readonly clock: Clock) {
        super();

        this.tool = gui_store.tool;
        this.internet_time_interval = setInterval(this.set_internet_time, 1000);
        this.set_internet_time();
    }

    dispose(): void {
        super.dispose();
        clearInterval(this.internet_time_interval);
    }

    set_tool(tool: GuiTool): void {
        this.gui_store.set_tool(tool);
    }

    private set_internet_time = (): void => {
        const now = this.clock.now();
        const s = now.getUTCSeconds();
        const m = now.getUTCMinutes();
        const h = (now.getUTCHours() + 1) % 24; // Internet time is calculated from UTC+01:00.
        this._internet_time.val = `@${Math.floor((s + 60 * (m + 60 * h)) / 86.4)}`;
    };
}
