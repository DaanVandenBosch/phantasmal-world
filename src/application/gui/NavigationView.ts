import { el } from "../../core/gui/dom";
import "./NavigationView.css";
import { gui_store, GuiTool } from "../../core/stores/GuiStore";
import { Widget } from "../../core/gui/Widget";
import { NavigationButton } from "./NavigationButton";

const TOOLS: [GuiTool, string][] = [
    [GuiTool.Viewer, "Viewer"],
    [GuiTool.QuestEditor, "Quest Editor"],
    [GuiTool.HuntOptimizer, "Hunt Optimizer"],
];

export class NavigationView extends Widget {
    readonly height = 30;

    private buttons = new Map<GuiTool, NavigationButton>(
        TOOLS.map(([value, text]) => [value, this.disposable(new NavigationButton(value, text))]),
    );

    constructor() {
        super(el.div({ class: "application_NavigationView" }));

        this.element.style.height = `${this.height}px`;
        this.element.onmousedown = this.mousedown;

        for (const button of this.buttons.values()) {
            this.element.append(button.element);
        }

        this.mark_tool_button(gui_store.tool.val);
        this.disposable(gui_store.tool.observe(({ value }) => this.mark_tool_button(value)));
    }

    private mousedown(e: MouseEvent): void {
        if (e.target instanceof HTMLLabelElement && e.target.control instanceof HTMLInputElement) {
            gui_store.tool.val = (GuiTool as any)[e.target.control.value];
        }
    }

    private mark_tool_button = (tool: GuiTool) => {
        const button = this.buttons.get(tool);
        if (button) button.checked = true;
    };
}
