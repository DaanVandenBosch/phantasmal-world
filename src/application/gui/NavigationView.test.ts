import { NavigationView } from "./NavigationView";
import { NavigationController } from "../controllers/NavigationController";
import { GuiStore } from "../../core/stores/GuiStore";
import { StubClock } from "../../../test/src/core/StubClock";

test("Should render correctly.", () => {
    const view = new NavigationView(
        new NavigationController(new GuiStore(), new StubClock(new Date("2020-01-01T00:30:01Z"))),
    );

    expect(view.element).toMatchSnapshot(
        "It should render a button per tool, the selected server, internet time and a github link icon.",
    );
});
