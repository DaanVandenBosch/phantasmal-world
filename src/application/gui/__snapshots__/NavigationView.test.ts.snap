// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly.: It should render a button per tool, the selected server, internet time and a github link icon. 1`] = `
<div
  class="application_NavigationView"
  style="height: 30px;"
>
  <span
    class="application_NavigationButton"
  >
    <input
      id="application_NavigationButton_Viewer"
      name="application_NavigationButton"
      type="radio"
      value="Viewer"
    />
    <label
      for="application_NavigationButton_Viewer"
    >
      Viewer
    </label>
  </span>
  <span
    class="application_NavigationButton"
  >
    <input
      id="application_NavigationButton_QuestEditor"
      name="application_NavigationButton"
      type="radio"
      value="QuestEditor"
    />
    <label
      for="application_NavigationButton_QuestEditor"
    >
      Quest Editor
    </label>
  </span>
  <span
    class="application_NavigationButton"
  >
    <input
      id="application_NavigationButton_HuntOptimizer"
      name="application_NavigationButton"
      type="radio"
      value="HuntOptimizer"
    />
    <label
      for="application_NavigationButton_HuntOptimizer"
    >
      Hunt Optimizer
    </label>
  </span>
  <div
    class="application_NavigationView_spacer"
  />
  <span
    class="application_NavigationView_server"
  >
    <label
      class="core_Label disabled"
      for="core_LabelledControl_id_0"
      title="Only Ephinea is supported at the moment"
    >
      Server:
    </label>
    <div
      class="core_Select disabled"
      id="core_LabelledControl_id_0"
      title="Only Ephinea is supported at the moment"
    >
      <button
        class="core_Button disabled"
        disabled=""
      >
        <span
          class="core_Button_inner"
        >
          <span
            class="core_Button_center"
          >
            Ephinea
          </span>
          <span
            class="core_Button_right"
          >
            <span>
              <span
                class="fas fa-caret-down"
              />
            </span>
          </span>
        </span>
      </button>
      <div
        class="core_Menu"
        hidden=""
        tabindex="-1"
      >
        <div
          class="core_Menu_inner"
        >
          <div
            data-index="0"
          >
            Ephinea
          </div>
        </div>
      </div>
    </div>
  </span>
  <span
    class="application_NavigationView_time"
    title="Internet time in beats"
  >
    @62
  </span>
  <a
    class="application_NavigationView_github"
    href="https://github.com/DaanVandenBosch/phantasmal-world"
    rel="noopener noreferrer"
    target="_blank"
    title="Phantasmal World is open source, code available on GitHub"
  >
    <span>
      <span
        class="fab fa-github"
      />
    </span>
  </a>
</div>
`;
